river/InputConfig.zig

```diff
@@ -0,0 +1,474 @@
+// This file is part of river, a dynamic tiling wayland compositor.
+//
+// Copyright 2021 The River Developers
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+const Self = @This();
+
+const build_options = @import("build_options");
+const std = @import("std");
+const wlr = @import("wlroots");
+
+const log = std.log.scoped(.input_config);
+
+const c = @import("c.zig");
+
+const server = &@import("main.zig").server;
+const util = @import("util.zig");
+
+const CommandError = @import("command.zig").Error;
+const InputDevice = @import("InputManager.zig").InputDevice;
+
+// TODO - keyboards
+//      - mapping to output / region
+//      - calibration matrices
+//      - scroll factor
+
+pub const EventState = enum {
+    unset,
+    enabled,
+    disabled,
+    disabled_on_external_mouse,
+
+    pub fn parse(str: []const u8) !EventState {
+        if (std.mem.eql(u8, str, "enabled")) {
+            return .enabled;
+        } else if (std.mem.eql(u8, str, "disabled")) {
+            return .disabled;
+        } else if (std.mem.eql(u8, str, "disabled-on-external-mouse")) {
+            return .disabled_on_external_mouse;
+        } else {
+            return CommandError.UnknownOption;
+        }
+    }
+
+    pub fn apply(event_state: EventState, device: *c.libinput_device) void {
+        const want = switch (event_state) {
+            .unset => return,
+            .enabled => c.LIBINPUT_CONFIG_SEND_EVENTS_ENABLED,
+            .disabled => c.LIBINPUT_CONFIG_SEND_EVENTS_DISABLED,
+            .disabled_on_external_mouse => c.LIBINPUT_CONFIG_SEND_EVENTS_DISABLED_ON_EXTERNAL_MOUSE,
+        };
+        const current = c.libinput_device_config_send_events_get_mode(device);
+        if (want != current)
+            _ = c.libinput_device_config_send_events_set_mode(device, @intCast(u32, want));
```

Comment:
Not a hard rule since I don't have any way to enforce it, but the style I've been moving towards for if statements is no braces if they fit on one line but always braces if they don't:
```zig
if (foo) bar(); // good

if (foo) {
    bar(); // good
}

if (foo)
    bar(); // I don't like this
```